core/controller/src/main/scala/whisk/core/loadBalancer/DistributedLoadBalancerData.scala

```diff
@@ -17,69 +17,102 @@
+class DistributedLoadBalancerData(instance: InstanceId, monitor: Option[ActorRef] = None)(
+  implicit actorSystem: ActorSystem,
+  logging: Logging)
+    extends LoadBalancerData {
 
   implicit val timeout = Timeout(5.seconds)
   implicit val executionContext = actorSystem.dispatcher
-  private val activationsById = TrieMap[ActivationId, ActivationEntry]()
+  private val overflowKey = "overflow"
+  private val activationsById = mutable.Map[ActivationId, ActivationEntry]()
+
+  private val localData = new LocalLoadBalancerData()
+  private var sharedDataInvokers = Map[String, Map[Int, Int]]()
+  private var sharedDataNamespaces = Map[String, Map[Int, Int]]()
+  private var sharedDataOverflow = Map[String, Map[Int, Int]]()
+
+  private val updateMonitor = actorSystem.actorOf(Props(new Actor {
+    override def receive = {
+      case Updated(storageName, entries) =>
+        monitor.foreach(_ ! Updated(storageName, entries))
+        storageName match {
+          case "Invokers"   => sharedDataInvokers = entries
+          case "Namespaces" => sharedDataNamespaces = entries
+          case "Overflow"   => sharedDataOverflow = entries
+        }
+    }
+  }))
 
   private val sharedStateInvokers = actorSystem.actorOf(
-    SharedDataService.props("Invokers"),
+    SharedDataService.props("Invokers", updateMonitor),
     name =
       "SharedDataServiceInvokers" + UUID())
   private val sharedStateNamespaces = actorSystem.actorOf(
-    SharedDataService.props("Namespaces"),
+    SharedDataService.props("Namespaces", updateMonitor),
     name =
       "SharedDataServiceNamespaces" + UUID())
-
-  def totalActivationCount =
-    (sharedStateInvokers ? GetMap).mapTo[Map[String, BigInt]].map(_.values.sum.toInt)
-
-  def activationCountOn(namespace: UUID): Future[Int] = {
-    (sharedStateNamespaces ? GetMap)
-      .mapTo[Map[String, BigInt]]
-      .map(_.mapValues(_.toInt).getOrElse(namespace.toString, 0))
+  private val sharedStateOverflow = actorSystem.actorOf(
+    SharedDataService.props("Overflow", updateMonitor),
+    name =
+      "SharedDataServiceOverflow" + UUID())
+  def totalActivationCount = {
+    val shared = sharedDataInvokers.values.flatten.filter(_._1 != instance.toInt).map(_._2).sum
+    shared + localData.totalActivationCount
+  }
+  def activationCountOn(namespace: UUID): Int = {
+    val shared = sharedDataNamespaces.getOrElse(namespace.toString, Map()).filter(_._1 != instance.toInt).values.sum
+    shared + localData.activationCountOn(namespace)
   }
 
-  def activationCountPerInvoker: Future[Map[String, Int]] = {
-    (sharedStateInvokers ? GetMap).mapTo[Map[String, BigInt]].map(_.mapValues(_.toInt))
+  def activationCountPerInvoker: Map[String, Int] = {
+    val shared = sharedDataInvokers.mapValues(_.filter(_._1 != instance.toInt).values.sum)
+    val local = localData.activationCountPerInvoker
+    local ++ shared.map { case (k, v) => k -> (v + local.getOrElse(k, 0)) }
```

Comment:
My understanding is that this merges local + shared state on **each** request. Can we rather merge updates into the local state and have each request be just a `get`?

This comment is valid for all "requests"

Reply:
I'm not sure - since the "shared" data is async and the local data is not, we need to assemble the data somehow to get an accurate map of values. 

For example, if we just use "shared" store to manage everything, that may be enough, but I'm not sure that gets accurate values (but maybe that is ok). I'll try it.

Reply:
Latest change (7b7f600) attempts this by a) extending LocalLoadBalancerData and b) resetting state periodically when replication updates are received. 

I need to test this more, and I'm not sure if there are better ways to do the update (currently does a clear() + foreach() to rebuild the TrieMaps) but let me know if you have thoughts on this approach.